### ISSUE TITLE
Formalize top-level ES exports

### ISSUE THREAD

**Issue by gaearon** [facebook/react#11503]
Currently we only ship CommonJS versions of all packages. However we might want to ship them as ESM in the future (https://github.com/facebook/react/issues/10021).

We can't quite easily do this because we haven't really decided on what top-level ES exports would look like from each package. For example, does `react` have a bunch of named exports, but also a default export called `React`? Should we encourage people to `import *` for better tree shaking? What about `react-test-renderer/shallow` that currently exports a class (and thus would start failing in Node were it converted to be a default export)?

**Comment by Andarist** [comment]
Imho `import *` is a way to go, Im not opposed to having a default export too, but it shouldnt be used to reexport other stuff like in this example:
```js
export const Component = ...
export default React
React.Component = Component
```

**Comment by gaearon** [comment]
>but it shouldnt be used to reexport other stuff like in this example:

Is there a technical reason why? (Aside from having two ways to do the same thing.)

My impression is that people who would import `*` (and not use the default) wouldn't have problems tree shaking since default would stay unused. But maybe I overestimate Rollup etc.

**Comment by Andarist** [comment]
That questions can be probably best answered by @lukastaegert. Ain't sure if something has changed since https://github.com/facebook/react/issues/10021#issuecomment-335128611

Also Rollup is not the only tree shaker out there, and while webpack's tree-shaking algorithm is worse than the one in rollup, it's usage is probably way higher than rollup's (both tools do excellent jobs ofc, I don't want to offend anyone, just stating facts) and if we can (as the community) help both tools at once we should do so whenever we can.

**Comment by jquense** [comment]
is tree-shaking going to _do_ anything in React's case, given that everything is preprocessed into a single flat bundle? I wonder what the primary import style is for React, personally i tend to treat it like a default export e.g. `React.Component`, `React.Children` but occasionally do the named thing with `cloneElement`

**Comment by lukastaegert** [comment]
As @gaearon already stated elsewhere, size improvements in case of react are expected to be minimal. Nevertheless, there ARE advantages:
* React.Children might probably be removed in some cases (so I heard 😉)
* React itself can be hoisted into the top scope by module bundlers that support this. This could again remove quite a few bytes and might also grant an oh-so-slight performance improvement. The main improvement would lie in the fact that there does not need to be another variable that references `React.Component` for every module but just one that is shared everywhere (this is how rollup usually does it). Also, though this is just me guessing, this might reduce the chance of webpack's ModuleConcatenationPlugin bailing out
* Static analysis for react is easier not only for module bundlers but also for e.g. IDEs and other tools. Many such tools already do a reasonable job at this for CJS modules but in the end, there is a lot of guessing involved on their side. With ES6 modules, analysis is a no-brainer.

As for the kind of exports, of course only named exports provide really provide the benefit of easy tree-shaking (unless you use GCC which might be able to do a little more in its aggressive move and maybe the latest rollup if you are really lucky). The question if you provide a default export as well is more difficult to decide:
* PRO: Painless migration for existing ES6 code bases (e.g. what @jquense describes)
* CON: Since everything is attached to a common object, once this object is included, all its keys are included at once which again defeats any attempts at tree-shaking. Even GCC might have a hard time here.

As a two-version migration strategy, you might add a default export in the next version for compatibility purposes which is declared deprecated (it might even display a warning via a getter etc.) and then remove it in a later version.

**Comment by gaearon** [comment]
This is also an interesting case: https://github.com/facebook/react/issues/11526. While monkeypatching for testing is a bit shady, we'll want to be conscious about breaking this (or having a workaround for it).

**Comment by Rich-Harris** [comment]
Came here via [this Twitter conversation](https://twitter.com/dan_abramov/status/933754555124322305). For me, there's a clear correct answer to this question: React and ReactDOM should **only** export named exports. They're not objects that contain state, or that other libraries can mutate or attach properties to (#11526 notwithstanding) — the only reason they exist is as a place to 'put' `Component`, `createElement` and so on. In other words, namespaces, which should be imported as such.

(It also makes life easier for bundlers, but that's neither here nor there.)

Of course, that does present a breaking change for people currently using a default import and transpiling. @lukastaegert probably has the right idea here, using accessors to print deprecation warnings. These could be removed in version 17, perhaps?

I don't have a ready-made suggestion for #11526 though. Perhaps shipping ESM would have wait for v17 for that reason anyway, in which case there'd be no need to worry about deprecation warnings.

**Comment by gaearon** [comment]
People have really come to like

```js
import React, { Component } from 'react'
```

so convincing them to give it up might be difficult.

I guess this is not too bad, even if a bit odd:

```js
import * as React from 'react';
import { Component } from 'react';
```

To clarify, we need `React` to be in scope (in this case, as a namespace) because JSX transpiles to `React.createElement()`. We could break JSX and say it depends on global `jsx()` function instead. Then imports would look like:

```js
import {jsx, Component} from 'react';
```

which is maybe okay but a huge change. This would also mean React UMD builds now need to set `window.jsx` too.

Why am I suggesting `jsx` instead of `createElement`? Well, `createElement` is already overloaded (`document.createElement`) and while it's okay with `React.` qualifier, without it claiming it on the global is just too much.

**Comment by Rich-Harris** [comment]
Confession: I always found it slightly odd that you have to explicitly import `React` in order to use JSX, even though you're not actually using that identifier anywhere. Perhaps in future, transpilers could insert `import * as React from 'react'` (configurable for the sake of Preact etc) on encountering JSX, if it doesn't already exist? That way you'd only need to do this...

```js
import { Component } from 'react';
```

...and the namespace import would be taken care of automatically.

**Comment by gaearon** [comment]
In a distant future, maybe. For now we need to make sure transpilers work with other module systems (CommonJS or globals). Making this configurable is also a hurdle, and further splits the community.

**Comment by Andarist** [comment]
What @Rich-Harris suggested (inserting a specific import when jsx is used) is easily done by transpilers plugin. The community would have to upgrade their `babel-plugin-transform-react-jsx` and that's it. And of course even existing setups would still work if only one adds `import * as React from 'react';` to the file.

Of course we need to consider other module systems, but it doesn't seem like a hard problem to solve. Are there any specific gotchas in mind?

**Comment by gaearon** [comment]
>Of course we need to consider other module systems, but it doesn't seem like a hard problem to solve. Are there any specific gotchas in mind?

I don’t know, what is your specific suggestion as to how to handle it? Would what the default be for Babel JSX plugin?

**Comment by jamiewinder** [comment]
> People have really come to like
>
>```
>import React, { Component } from 'react'
>```

What people? Come forth so that I may mock thee.

**Comment by gaearon** [comment]
I did that a lot 🙂 Pretty sure I've seen this in other places too.

**Comment by Andarist** [comment]
Default is at the moment [`React.createElement`](`https://github.com/babel/babel/tree/6d820a27578254603ad6d6623277f31d5410a37c/packages/babel-plugin-transform-react-jsx#pragma`) and it would pretty much stay the same. The only problem is that it assumes a global now (or already available in the scope). 

I think as es modules are basically the standard way (although not yet adopted by all) of doing modules, it is reasonable to assume majority is (or should) use it. Vast majority already uses various build step tools to create their bundles - which is even more true in this discussion because we are talking about **transpiling jsx syntax**. Changing the default behaviour of the jsx plugin to auto insertion of `React.createElement` into the scope is imho reasonable thing to do. We are at the perfect time for this change with babel@7 coming soon (-ish). With recent addition of [`babel-helper-module-imports`](https://github.com/babel/babel/tree/6d820a27578254603ad6d6623277f31d5410a37c/packages/babel-helper-module-imports) it is also easier than ever to insert the right type of the import (es/cjs) to the file.

Having this configurable to bail out to today's behaviour (assuming present in scope) seems really like a minor change in configuration needed for a minority of users and an improvement (sure, not a big one - but still) for majority.

**Comment by kzc** [comment]
> Should we encourage people to import * for better tree shaking? 

Thanks to @alexlamsl `uglify-es` has eliminated the `export default` penalty in common scenarios:

```js
$ cat mod.js 
export default {
	foo: 1,
	bar: 2,
	square: (x) => x * x,
	cube: (x) => x * x * x,
};
```
```js
$ cat main.js 
import mod from './mod.js'
console.log(mod.foo, mod.cube(mod.bar));
```
```js
$ rollup main.js -f es --silent | tee bundle.js
var mod = {
	foo: 1,
	bar: 2,
	square: (x) => x * x,
	cube: (x) => x * x * x,
};

console.log(mod.foo, mod.cube(mod.bar));
```
```
$ uglifyjs -V
uglify-es 3.2.1
```
```js
$ cat bundle.js | uglifyjs --toplevel -bc
var mod_foo = 1, mod_bar = 2, mod_cube = x => x * x * x;

console.log(mod_foo, mod_cube(mod_bar));
```
```js
$ cat bundle.js | uglifyjs --toplevel -mc passes=3
console.log(1,8);
```

**Comment by Andarist** [comment]
wow, that's great new 👏 is `uglify-es` considered to be stable now? I recall you mentioning few months back that it isn't there quite yet, but I can remember that incorrectly, so ain't sure.

Anyway - that's all and nice in a rollup world, but considering that `React` is bundled mostly in apps and those use mostly `webpack` which does not do scope hoisting by default, I'd still say that exporting an object as default should be avoided to aid other tools than `uglisy-es`+`rollup` in their efforts to produce smaller bundle sizes. Also for me it is semantically better to avoid this - what libs actually do in such cases is providing a namespace and it is better represented when using `import * as Namespace from 'namespace'`

**Comment by kzc** [comment]
> is uglify-es considered to be stable now?

As stable as anything else in the JS ecosystem. Over 500K downloads per week. 

> that's all and nice in a rollup world, but considering that React is bundled mostly in apps and those use mostly webpack which does not do scope hoisting by default

Anyway, it's an option. Webpack defaults are not ideal anyway - you have to use `ModuleConcatenationPlugin` as you know.

**Comment by lukastaegert** [comment]
Adding a few cents here:
* I totally agree with @Rich-Harris that semantically, named exports are the right choice
* I really do not like either `import React from 'react'` or `import * as React from 'react'` just to be able to use JSX syntax. In my eyes, this design is clearly violating the Interface Segregation Principle in that it forces users to import all of React just to be able to use the `createElement` part.

So if we are at a point where we might make breaking-change decisions, I would advise to change this so that JSX depends on a single (global or imported) function. I would have called it `createJSXElement()`, which in my opinion describes it even better than `createElement()` and no longer needs the React context to make sense. But in a world where every byte counts, `jsx()` is probably ok, too.

This would also at last decouple JSX from React in a way such that other libraries can choose to support JSX by using the same transformation and supplying a different `jsx` function. Of course you have a lot of responsibility here guiding countless established applications through such a transformation but from an architectural point of view, this is where I think React and JSX should be heading. Using Babel to do the heavy lifting of such a transformation sounds like a great idea to me!

**Comment by Andarist** [comment]
Personally I do not see much gain in migrating to `jsx` helper as the default IMHO for the babel plugin should be importing it from the `react` package, so the name of the actual helper doesn't really matter - the rest is just matter of having it configurable.

**Comment by NMinhNguyen** [comment]
This is probably slightly tangential to the main discussion, but I'm curious how well ES modules work with conditionally checking `process.env.NODE_ENV` to export dev/prod bundles? For example,

https://github.com/facebook/react/blob/d9c1dbd61772f8f8ab0cdf389e70463d704c480b/packages/react/npm/index.js#L3-L7

I may be missing something obvious here, but I'm struggling to see how to translate this pattern into ES modules?

**Comment by milesj** [comment]
@NMinhNguyen Conditional exports aren't possible with ES modules.

**Comment by Andarist** [comment]
`process.env.NODE_ENV` checks can be at more granular (code) level though, ready to be replaced by the bundler with appropriate values.

**Comment by NMinhNguyen** [comment]
@Andarist @milesj Thanks for confirming my suspicion :)

> `process.env.NODE_ENV` checks can be at more granular (code) level though, ready to be replaced by the bundler with appropriate values.

From the React 16 [blog post](https://reactjs.org/blog/2017/09/26/react-v16.0.html#better-server-side-rendering) I thought that the `process.env.NODE_ENV` checks were pulled out to the very top on purpose (as opposed to them being more granular, which is what they are in the source, if I'm not mistaken), to help performance in Node.js?

> ### Better server-side rendering
>
> React 16 includes a completely rewritten server renderer. It's really fast. It supports **streaming**, so you can start sending bytes to the client faster. And thanks to a [new packaging strategy](#reduced-file-size) that compiles away `process.env` checks (Believe it or not, reading `process.env` in Node is really slow!), you no longer need to bundle React to get good server-rendering performance.

Like, I'm not sure how one could use the `module` field in `package.json` and differentiate between dev/prod for ESM.

**Comment by Andarist** [comment]
> Like, I'm not sure how one could use the module field in package.json and differentiate between dev/prod for ESM while keeping ES bundles flat and not affecting Node.js perf

This for sure is a drawback, because there is no standard way at the moment for doing this. OTOH it's just a matter of tooling, it is possible (and it's rather easy) to compile this in build steps of your application even today. Ofc it would be easier if package could expose dev/prod builds and the resolver would just know which one to pick, but maybe that's just a matter of pushing this idea to tooling authors.

**Comment by chyzwar** [comment]
For class:
```js
import Component from 'react/Component'

class MyButton extends Component{
  constructor(){
    this.state = {}
  }

  render() {
    return <button> Button <Button>
  }
}
```
Where transform will use super.createElement() to transform to jsx or use static Component.createElement(). 

For stateless components:

```js
import jsx from 'react/jsx'

const MyButton = () => jsx`<button> Button <Button>`;
```
it is maybe possible to use tagged template literal?

Node hopefully accept this PR https://github.com/nodejs/node/pull/18392

**Comment by TheLarkInn** [comment]
We agree here with @Rich-Harris.

**Comment by jimniels** [comment]
Just dropping a comment on this thread which hasn't really been mentioned specifically.

I’m in a situation where I’m not using a bundler at all and just want to import react and various components for use natively through the browser (`<script type="module" src="...">`), i.e.

```js
import React from “https://unpkg.com/react@16.4.1/umd/react.development.js”;
import ReactDOM from “https://unpkg.com/react-dom@16.4.1/umd/react-dom.development.js”;
ReactDOM.render(
  React.createElement(...),
  document.getElementById('root')
);
```

From what I can tell, this isn't possible today. Instead, I have to include the UMD version of react via a `<script>` tag from the CDN and then assume it’s presence on the window in any `<script type="module">` module I write:

```
// myPage.html
<div id="myComponentRoot"></div>
<script src="https://unpkg.com/react@16.4.1/umd/react.development.js"></script>
<script type="module" src="/assets/scripts/components/MyComponent.js"></script>

// MyComponent.js
import AnotherComponent from "/assets/scripts/components/AnotherComponent.js";
window.ReactDOM.render(
  window.React.createElement(AnotherComponent),
  document.getElementById('root')
);

// AnotherComponent.js
export default class AnotherComponent extends window.React.Component {...}
```

Having a react import from a CDN would be fantastic. It would make prototyping in the browser very quick and easy while still being able to maintain separation of files. One thing I always felt I was sacrificing when using React without a bundler was the ability to separate components (and other utility functions, etc) by file. But now with browser support for native ES modules, I can write my React components in separate files and have the browser just consume them as they're written. Granted that's if I'm not using JSX, but even if I was using JSX, I could transpire all the files in place via a build step and all my imports would still work in the browser.

```js
// /assets/scripts/entry.js
import React from “https://unpkg.com/react@16.4.1/umd/react.development.js”;
import React from “https://unpkg.com/react-dom@16.4.1/umd/react-dom.development.js”;
import RelatedPosts from "/assets/scripts/components/RelatedPosts.js";
ReactDOM.render(
  React.createElement(RelatedPosts),
  document.getElementById('root')
);

// /assets/scripts/components/RelatedPosts.js
import React from “https://unpkg.com/react@16.4.1/umd/react.development.js”;
import ListItem from "/assets/scripts/components/ListItem.js"
export default class MyComponent extends React.Component {
  componentDidMount() { /* fetch some data */ }
  render() { 
    return React.createElement(
      'ul',
      {},
      this.state.items.map(item => React.createElement(ListItem, { item: item })
    )
  }
}

// /assets/scripts/components/ListItem.js
import React from “https://unpkg.com/react@16.4.1/umd/react.development.js”;
export default function ListItem(props) {
  return React.createElement('li', null, ...)
}
```

I'm sure some people would argue typing that CDN url all the time is a problem (a problem some people are trying to fix) but the tradeoffs are worth it to me. Changing/updating that url is a simple find/replace. For my use case, this outweighs the trouble of setting up a bundler. 

If React had support for something like this, there would be no need for tooling. I'm just using the browser. I could ship code like this in a few personal projects which assume modern browsers and use react as a progressive enhancement on the page. What makes this fantastic is when I come back to the code base in 12 months, I don't have to change a bunch of tooling APIs or even have NPM as a package manager. I'm just using APIs from the browser, nothing else.

FWIW: if/when React does ship with support like this, I think it could be very valuable to show how you could use React like this in the docs, teaching that you *can* leverage React and its component model by separating each components logic via its own file and you don't need a bundler to do it, just use native `<script type="module">`, import React from a CDN (or your own local copy), and your off!”

**Comment by kenokabe** [comment]
Now, all of the modern browsers including mobile versions support ESM.  ESM is no longer a future module system but a current defacto-standard.

Please be aware of not providing the standardized module is a critical problem, especially for a defacto-standard web library.

**Comment by TrySound** [comment]
I started working on this [here](https://github.com/facebook/react/pull/13321) and [here](https://github.com/facebook/react/pull/13080)

**Comment by kenokabe** [comment]
@TrySound  Thanks for your contribution.
Is there any place to grab and test the ESM build?

**Comment by TrySound** [comment]
It's ready only for react-is package.

**Comment by kenokabe** [comment]
@TrySound 
Ok, I found your branch https://github.com/TrySound/react/tree/react-is-esm, and have built, and now I know what you meant. Looking forward to `react-dom` too.

**Comment by PM5544** [comment]
@kenokabe some time ago (awaiting resolution of these issues) I threw together a little thing to output an ESM version of react and react-dom. It's [on NPM](https://www.npmjs.com/package/react-ecmascript)

**Comment by kenokabe** [comment]
I think the React community discussed this issue quite enough for a while.
https://discuss.reactjs.org/t/es6-import-as-react-vs-import-react/360/9

Please decide the official r ES6 module specification, and publish soon.

**Comment by TrySound** [comment]
@kenokabe We are on the way. Don't force us please. It's not that easy.

**Comment by kenokabe** [comment]
@TrySound My apologies.
  I didn't mean to you, since the head mention of this topic is

>We can't quite easily do this because we haven't really decided on what top-level ES exports would look like from each package. For example, does react have a bunch of named exports, but also a default export called React? Should we encourage people to import * for better tree shaking? 

and, it' while ago, and I just thought it's been discussed in React community, so I wanted to suggest the decision would be clear. Thanks!

**Comment by leoyli** [comment]
Want get some update on this...

I'm using webpack v4 for bundling our application, while my IDE intellisense (WebStorm) suggest me to use `import * as React from 'react';` while my coworker ask me to change `import React from 'react';` in a code review. Both works fine so I thought he is saying some nonsense, but to make him happy I'm changing it anyway. That's also how I find this thread.

While out of curious, I compare the differences at the final build size between it (with React 16.8.1):

In `import * as React from 'react';`: 6,618,723 bytes
In `import React from 'react';`: 6,619,077 bytes

So obviously, it did have some differences, marginal though. (note. the same with `propTypes`)

If my understanding in this thread correctly, it would be favour to have `import * as React from 'react';` actually, right?! Because (1) yes, it did save some size; (2) ESM is a scandalized way so no more CJS leftovers. If that is the case I'd like to change this today and align with my IDE.

**Comment by TrySound** [comment]
@leoyli In long term yes. But first there will be both named and default exports to not break existing code.

**Comment by lukejacksonn** [comment]
I took matters into my own hands here, somewhat as an experiment as I am not using a bundler in my project anymore and wanted to still use react (direct from unpkg.com like you can with other libraries such as Vue, Hyperapp etc.). This is what I came up with:

https://github.com/lukejacksonn/es-react
> An ES6 module exposing the latest version of `React` and `ReactDOM`

As described in the README this is mostly a POC but for people that cannot wait for this build to land then it is a `16.8.3` build which includes hooks, suspense, lazy etc. and works as expected by doing: 

```js
import { React, ReactDOM } from 'https://unpkg.com/es-react'
```

Maybe some of you in this thread will find it useful.. personally I have been using this approach to create a [build step free react starter project](https://github.com/lukejacksonn/create-es-react-app). It is also still a work in progress.

**Comment by PM5544** [comment]
@lukejacksonn We've been using such a solution on production as well, while our approach was different in the sense that it is more a transformer script for the UMD version of React and ReactDOM in your current project. And that it outputs these file separately so for the most code out there it should be a drop in replacement. If you're interested https://github.com/wearespindle/react-ecmascript and you can also load it from unpkg https://unpkg.com/react-ecmascript/

**Comment by lukejacksonn** [comment]
@PM5544 Oh wow.. this is a much more comprehensive solution than mine! Great job 💯

**Comment by ghengeveld** [comment]
Awesome stuff @PM5544. Would love to hear more about it someday. Maybe a guest appearance back at Xebia? 
I've recently adopted [pack](https://github.com/pikapkg/pack) to bundle my open source packages, which supports UNPKG.
Anyone know of a good article on loading dependencies from UNPKG directly rather than using a bundler?

**Comment by lukejacksonn** [comment]
I'm currently writing one and will be giving it as a talk at React Norway in June too!

**Comment by sdegutis** [comment]
@TrySound Has there been any update on this since February? What's left to get this issue moving, and can I participate in getting this issue closed by some coding work? I already signed the CA, and I have time available today to work on it.

**Comment by TrySound** [comment]
This need to be merged first https://github.com/facebook/react/pull/15037

**Comment by sdegutis** [comment]
@TrySound Okay thanks I have forwarded my offer for assistance to that thread.

**Comment by sokra** [comment]
When you go with a default React export you can go with this approach:

``` js
// react/index.js
import * as React from "./react";
export { React as default }
export * from "./react";

// react/react.js
export function createElement() {}
...
```

This make it statically analyse-able that the default export is a namespace object which allows tree-shaking for these constructs in webpack 5:

``` js
import React from "react";

React.createElement(); // <- only `createElement` export is used
```

**Comment by lukastaegert** [comment]
Rollup [would love that too](https://rollupjs.org/repl/?version=1.17.0&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMFJlYWN0JTIwZnJvbSUyMCcuJTJGaW5kZXgnJTNCJTVDbmltcG9ydCUyMCcuJTJGZGVwJyUzQiU1Q25pbXBvcnQlMjAnLiUyRmRlcDInJTNCJTVDbiU1Q24lMkYlMkYlMjB2aWElMjBkZWZhdWx0JTIwaW1wb3J0JTVDbmNvbnNvbGUubG9nKFJlYWN0LmNyZWF0ZUVsZW1lbnQoKSklMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSU3RCUyQyU3QiUyMm5hbWUlMjIlM0ElMjJpbmRleC5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjJpbXBvcnQlMjAqJTIwYXMlMjBSZWFjdCUyMGZyb20lMjAnLiUyRnJlYWN0JyUzQiU1Q25leHBvcnQlMjAlN0IlMjBSZWFjdCUyMGFzJTIwZGVmYXVsdCUyMCU3RCU1Q25leHBvcnQlMjAqJTIwZnJvbSUyMCcuJTJGcmVhY3QnJTNCJTIyJTJDJTIyaXNFbnRyeSUyMiUzQWZhbHNlJTdEJTJDJTdCJTIybmFtZSUyMiUzQSUyMnJlYWN0LmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmV4cG9ydCUyMGZ1bmN0aW9uJTIwY3JlYXRlRWxlbWVudCUyMCgpJTIwJTdCJTVDbiU1Q3RyZXR1cm4lMjAnaGVyZSUyQyUyMGVsZW1lbnRzJTIwYXJlJTIwY3JlYXRlZCclM0IlNUNuJTdEJTVDbiU1Q25leHBvcnQlMjBjbGFzcyUyMENvbXBvbmVudCU3QiU1Q24lNUN0Y29uc3RydWN0b3IoKSUyMCU3QiU1Q24lNUN0JTIwJTIwY29uc29sZS5sb2coJ25vdCUyMHVzZWQnKSUzQiU1Q24lNUN0JTdEJTVDbiU3RCUyMCUyMiUyQyUyMmlzRW50cnklMjIlM0FmYWxzZSU3RCUyQyU3QiUyMm5hbWUlMjIlM0ElMjJkZXAuanMlMjIlMkMlMjJjb2RlJTIyJTNBJTIyaW1wb3J0JTIwKiUyMGFzJTIwUmVhY3QlMjBmcm9tJTIwJy4lMkZpbmRleCclNUNuJTVDbiUyRiUyRiUyMHZpYSUyMHN0YXIlMjBpbXBvcnQlNUNuY29uc29sZS5sb2coUmVhY3QuY3JlYXRlRWxlbWVudCgpKSUyMiUyQyUyMmlzRW50cnklMjIlM0FmYWxzZSU3RCUyQyU3QiUyMm5hbWUlMjIlM0ElMjJkZXAyLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3QmNyZWF0ZUVsZW1lbnQlN0QlMjBmcm9tJTIwJy4lMkZpbmRleCclNUNuJTVDbiUyRiUyRiUyMHZpYSUyMG5hbWVkJTIwaW1wb3J0JTVDbmNvbnNvbGUubG9nKGNyZWF0ZUVsZW1lbnQoKSklMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXNtJTIyJTJDJTIybmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyYW1kJTIyJTNBJTdCJTIyaWQlMjIlM0ElMjIlMjIlN0QlMkMlMjJnbG9iYWxzJTIyJTNBJTdCJTdEJTdEJTJDJTIyZXhhbXBsZSUyMiUzQW51bGwlN0Q=)

**Comment by GrosSacASac** [comment]
I am in the rollup gitter chat for 1.5 years and this kind of issues comes up every 2 weeks or so ...

**Comment by lukastaegert** [comment]
BTW, @lukejacksonn did a tremendous job on the inofficial [es-react](https://www.npmjs.com/package/es-react) fork, can highly recommend it.

**Comment by stken2050** [comment]
> BTW, @lukejacksonn did a tremendous job on the inofficial [es-react](https://www.npmjs.com/package/es-react) fork, can highly recommend it.

I wonder why the official FB team does nothing to this.

**Comment by lukastaegert** [comment]
I was also hoping this would generate some pressure to move things forward, and I guess so was @lukejacksonn himself. As I understand, though, he actually received some support from the React team to build his fork from the original sources, so there seems to be at least some interest in this.

**Comment by lukejacksonn** [comment]
I hoped this too indeed. Actually I received next to no support from the react team in the creation of the package (besides some kind words of encouragement from @threepointone). Recently a colleague here at Formidable helped me build the package programatically which is an improvement over donig it by hand every time a new version of react is released but does result in much less clean output in the network tab so I am not sure if it will stay like this yet. We shall see!

**Comment by osener** [comment]
This is something that caused a ton of friction for me as well. In one project every bit of my code & dependencies used ES6 and worked well with Rollup, and I "solved" this by switching to Preact which is distributed as ES6 modules.

Otherwise, this tip works well with rollup: https://github.com/rollup/rollup-plugin-commonjs/issues/407#issuecomment-527837831

**Comment by foisonocean** [comment]
It's almost 2020 now, I would like to know if there are any updates from the official FB team? Would there be any changes related to this in React v17?

**Comment by tomByrer** [comment]
For those needing an updated ES-module React NOW, try `@pica/react`, which is already on v16.13.x
https://www.npmjs.com/package/@pika/react
https://www.npmjs.com/package/@pika/react-dom
https://github.com/pikapkg/react

**Comment by Falci** [comment]
> In a distant future, maybe.

Ok, already in the future. Is it in the near future now?

**Comment by wereHamster** [comment]
@gaearon what is the blocker for deciding how to structure the exports (default export vs named exports)? Is there anything the community can help you with making this decision?

**Comment by wereHamster** [comment]
Apparently that decision has already been made a while ago: #18102. This issue can be closed now.

**Comment by gaearon** [comment]
I'll give a small update on this.

## No Default Exports

Our eventual plan is to move away from default exports entirely:

```js
import { useState } from 'react';
```

In that world, this would **not** work:

```js
import React from 'react'; // no
```

This would work although it's a bit noisy:

```js
import * as React from 'react';
```

But here's the kicker. **There wouldn't actually be much reason to import `React` at all.**

## JSX Auto Import

The reason people import `React` today is mostly due to JSX. But @lunaruan is finishing work on the [new JSX transform](https://babeljs.io/blog/2020/03/16/7.9.0#a-new-jsx-transform-11154-https-githubcom-babel-babel-pull-11154) and related codemods, which remove the need for it.

So you would go from this:

```js
import React from 'react';
import { useState } from 'react';

function Button() {
  const [pressed, setPressed] = useState(false)
  return <button />
}
```

to this:

```js
import { useState } from 'react';

function Button() {
  const [pressed, setPressed] = useState(false)
  return <button />
}
```

JSX inserts the correct import automatically under the hood, so no need for `React` in scope.
This is what makes the move to remove default exports tolerable. You just don't need them as much.

## ES Modules

Deploying ESM beyond an overall small slice of enthusiasts is challenging. The broad ecosystem isn't really ready and there's a ton of ways in which things go wrong with different tool combinations. The way CJS and ESM interact is very complex, and that interop (and how it fails) is the source of most of these issues.

So our current thinking is that when we go ESM, we might want to try go ESM all the way. No CJS at all — or separated in a compat legacy package. This won't happen in React 17 and is unlikely in 18, but is plausible to try in React 19.

**Comment by joeldenning** [comment]
For anybody looking for an alternative to `@pika/react`'s ESM build that is usable in browsers without an import map polyfill, check out https://github.com/esm-bundle/react and https://github.com/esm-bundle/react-dom.

Code Sandbox demonstrating: https://codesandbox.io/s/gifted-roentgen-qcqoj?file=/index.html

**Comment by gaearon** [comment]
A few people have challenged the notion that the ecosystem isn't ready. I don't have full context on this but if you think this is a good time to start making changes, I'd appreciate if you could look over https://github.com/reactjs/rfcs/pull/38 and express any concerns about it. Is that roughly would you had in mind, or would you prefer a different approach?

**Comment by wereHamster** [comment]
> But here's the kicker. There wouldn't actually be much reason to import React at all.

There is if you are using TypeScript, and will continue so for the foreseeable future. Until TS learns about this new magic behaviour of babel, developers will have to continue to explicitly import React, and they need to know what the correct import statement is.

> JSX inserts the correct import automatically under the hood, so no need for React in scope.

s/JSX/The new babel react jsx transform plugin/. JSX is a syntax extension to JavaScript, it by itself doesn't do anything.

**Comment by gaearon** [comment]
>There is if you are using TypeScript, and will continue so for the foreseeable future. Until TS learns about this new magic behaviour of babel, developers will have to continue to explicitly import React, and they need to know what the correct import statement is.

The TypeScript team is aware of this change, and it's being tracked in https://github.com/microsoft/TypeScript/issues/34547. We are in close contact with them so please be assured it's not a second-class citizen for us.

>s/JSX/The new babel react jsx transform plugin/

Yes, this is what I meant!

**Comment by joeldenning** [comment]
> I'd appreciate if you could look over reactjs/rfcs#38 and express any concerns about it. Is that roughly would you had in mind, or would you prefer a different approach?

Some of the RFC's original text is out of date. NodeJS allows for ESM to run in `.js` files instead of `.mjs` now, when you specify a `"type"` in your package.json. ([docs](https://nodejs.org/api/esm.html#esm_enabling)).

Specifically, the RFC original text says the following which is not true:

> ESM code **must** be inside of a `.mjs` file

After talking with @frehner, here's our proposal for how ESM could be brought to React. Note that we do not couple the named/default exports problem with publishing an ESM version of React. @gaearon has clarified that the default export will eventually go away, but that is not listed in our proposal until Phase 4.

|  | Phase 1 | Phase 2 | Phase 3 | Phase 4 |
| - | -------- | -------- | -------- | ------- |
| ESM published? | ✔️ | ✔️ | ✔️ | ✔️ |
| [package.json `"module"`](https://github.com/webpack/webpack/issues/4674) | ❌ | ✔️ | ✔️ | ✔️ |
| webpack/rollup use esm <sup>[1](https://webpack.js.org/configuration/resolve/#resolvemainfields), [2](https://github.com/rollup/plugins/tree/master/packages/node-resolve#mainfields)</sup> | ❌ | ✔️ | ✔️ | ✔️ |
| [package.json `"exports"`](https://nodejs.org/api/esm.html#esm_exports_sugar) | ❌ | ❌ | ✔️ | ✔️ |
| [package.json `"type"`](https://nodejs.org/api/esm.html#esm_package_json_type_field) | ❌ | ❌ | ✔️  | ✔️ |
| [NodeJS uses esm](https://nodejs.org/api/esm.html#esm_enabling) | ❌ | ❌ | ✔️ | ✔️ |
| Breaking change? | ❌ | ❌ | ❓ | ✔️ |
| Default export gone? | ❌ | ❌ | ❌ | ✔️ |
| File extensions required in imports | ❌ | ❌ | ❌ | ❌ |
| mjs file extensions | ❌ | ❌ | ❌ | ❌ |

I think there is a valid argument for combining Phase 1 and Phase 2 together, since Phase 1 really only targets enthusiasts. However, I split them because I think that separating the phases gives a chance to very slowly roll out ESM in a way that won't immediately break CRA and the entire ecosystem without first giving the early adopters a chance to report issues and find any fixes.

**Comment by Andarist** [comment]
@joeldenning what would be the rough timing of the phases? is it just time-based or are those related to some time checkpoints in the ecosystem? Would `require('react')` still work in Phase 3?

**Comment by joeldenning** [comment]
> what would be the rough timing of the phases? is it just time-based or are those related to some time checkpoints in the ecosystem?

The timing for all of the phases is only limited by the work to be done and React's release schedule. All of the proposed things are supported by bundlers and nodejs, with appropriate fallbacks when using older versions of them.

> Would require('react') still work in Phase 3?

Yes, see https://nodejs.org/api/esm.html#esm_dual_commonjs_es_module_packages and https://nodejs.org/api/esm.html#esm_package_entry_points

**Comment by joeldenning** [comment]
An additional clarification, here is what we're proposing the published tarball would look like for React:

```
node_modules/react/
  cjs/
    react.development.js
    react.production.min.js
    react.profiling.min.js
  umd/
    react.development.js
    react.production.min.js
    react.profiling.min.js
  esm/
    react.development.js
    react.production.min.js
    react.profiling.min.js
  index.js
```

And here is an approximation of what the package.json would be at the end:

```json
{
  "type": "module",
  "main": "index.js",
  "module": "esm/react.development.js",
  "exports": {
    "import": "./esm/react.development.js",
    "require": "./cjs/react.development.js"
  }
}
```

^ This isn't fully thought through and perfected, but I'm sharing to give concrete context to what I'm proposing.

**Comment by Andarist** [comment]
So - React suffers from the dual package hazard because it's stateful (hooks), so this state would have to be carefully isolated. Either it should live in a small CJS file that could be imported by both CJS and ESM entries or maybe there is a way load `.json` and mutate it with that state from both entries (I'm not 100% sure about the second approach).

I also think that it's important to list in your table adding named exports, which would happen already in Phase 1. 

I'm still not 100% sure if I haven't missed some corner cases, this topic is very complex. At the same time, the schedule should be tied to the ecosystem checkpoints - Phase 3 can only be shipped once `exports` has sufficient support in bundlers, otherwise I feel this could lead to potential issues.

**Comment by joeldenning** [comment]
> React suffers from the dual package hazard because it's stateful (hooks), so this state would have to be carefully isolated.

Good point, I hadn't considered this ([more info](https://nodejs.org/api/esm.html#esm_dual_package_hazard)). In that case, perhaps your suggestion of a shared file between the CJS / ESM builds would be useful. Or perhaps the ESM version is not a full copy of react, but just the ESM public interface that calls into the CJS build.

> I also think that it's important to list in your table adding named exports, which would happen already in Phase 1.

Good point, I'll add that to the table 👍 

> I'm still not 100% sure if I haven't missed some corner cases, this topic is very complex. At the same time, the schedule should be tied to the ecosystem checkpoints - Phase 3 can only be shipped once exports has sufficient support in bundlers, otherwise I feel this could lead to potential issues.

Agreed about Phase 3 - this is why I put a question mark for whether it was a breaking change. I know that adding package.json exports has often been a breaking change for other packages in the ecosystem. And the topic is definitely complex. One thing to note is that the order of Phase 3 and 4 could be swapped, if desired. I think that those implementing each phase would have to do very very very thorough testing of many versions of webpack, rollup, nodejs, etc. I'm not saying that the work to be done is trivial - just saying that I do think that there is likely a transition pathway here :)

**Comment by Andarist** [comment]
> It seems that this is already the case, though, in the source code? From what I can tell, the source code already exports things as named exports.

Ah - right, in this case there should be a table row for adding `export default` 😂 as it currently works in most bundlers and is popular in the wild, but adding true ESM entry without providing `default` would break those usages.

**Comment by joeldenning** [comment]
> in this case there should be a table row for adding export default

Yes, good point. I'll add that. I think that a PR doing that would "seem bad," but I view it as just accepting the current public interface for what it is, with plans to improve in the future.

**Comment by gaearon** [comment]
We should note that as soon as *some* version is published, there's no going back and any semantic change to it would be breaking. Given we don't release majors very often, we should consider how to batch any changes so that there's least churn.

**Comment by gaearon** [comment]
There's also a question of how development/production build split would be handled. And indeed, stateful nature of the React build is pretty important to preserve.

**Comment by joeldenning** [comment]
> We should note that as soon as some version is published, there's no going back and any semantic change to it would be breaking. Given we don't release majors very often, we should consider how to batch any changes so that there's least churn.

Good points. My idea is to add an explicit `export default React` to an ESM build that is published in React 16. I think the PR might raise some eyebrows and might be controversial, as that is not the destination that has been decided on. However, my view is that we can have an ESM build in React 16 and then remove the `export default` in a future major version. To me, using react via `import React from 'react';` is so overwhelmingly common that exporting default in the ESM build is simply "accepting where we are." A future major version would remove it.

Also related to minimizing breaking changes - phase 3 and 4 could be part of the same major release. It's possible that phase 3 could be done in a fully backwards compatible way, in which case it could also sneak into a 16 release. But that one is trickier and I don't know enough about it to be confident in that.

> There's also a question of how development/production build split would be handled.

This is something I overlooked. I don't know how to do this with ESM in both bundlers and in NodeJS, but will do some research to see what's possible.

> And indeed, stateful nature of the React build is pretty important to preserve.

Agreed. One thing to note is that the stateful nature of the React build only needs to be solved in Phase 3, not in Phases 1 and 2. The options that @Andarist and I suggested would work for solving it.

**Comment by ljharb** [comment]
The easiest, most backwards-compatible approach *for now* is to add a simple ESM wrapper to allow for named imports.

I'd be happy to make a PR to add this to React, ensuring that the addition of the "exports" field is not a breaking change (i wrote a tool, `npx ls-exports`, which makes determining this easy). It won't help folks trying to use ESM without a build process, but that's a problem that individual packages aren't capable of solving anyways.

@gaearon would that be useful? It could land in React 16 as a semver-minor.

**Comment by gaearon** [comment]
>However, my view is that we can have an ESM build in React 16 and then remove the export default in a future major version.

Another concern is that increasing the number of configurations puts the strain on the ecosystem. E.g. I think if we do release an ESM build it should not include something that we're definitely going to remove in the very next release, such as a default export. In other words, I don't think it makes a lot of sense to introduce something that's going away (default exports) into something that's just being added (ESM build).

**Comment by frehner** [comment]
> > However, my view is that we can have an ESM build in React 16 and then remove the export default in a future major version.
> 
> Another concern is that increasing the number of configurations puts the strain on the ecosystem. E.g. I think if we do release an ESM build it should not include something that we're definitely going to remove in the very next release, such as a default export. In other words, I don't think it makes a lot of sense to introduce something that's going away (default exports) into something that's just being added (ESM build).

I think it depends on how long that next release is though. If we’re talking about a week difference, then that seems reasonable. However, if we’re talking months/years, then I don’t see why it would be bad to get it out for that amount of time

**Comment by gaearon** [comment]
The concern in this case is that the longer it's out there, the more people depend on it in the current form. And then introducing a breaking change there would make it more difficult for them to upgrade React. Now there is not just one ESM migration, but several, and some people will be left behind because they jumped too early and later don't have the resources for another migration. Which, in case of ESM, will ripple across the ecosystem.

**Comment by joeldenning** [comment]
> I think if we do release an ESM build it should not include something that we're definitely going to remove in the very next release, such as a default export. In other words, I don't think it makes a lot of sense to introduce something that's going away (default exports) into something that's just being added (ESM build).

I don't view `import React from 'react'` as something newly added, but rather an admission of current reality. Even though it was perhaps unintentional and only a side effect of now-obsolete ESM/CJS interop, there are still thousands (millions?) of lines of code that do it. The alternative of only exporting named exports says to users "we published an ESM build in a minor version, but you can't use it without changing all your code" which to me is more confusing for users than seeing it in major version's release notes.

**Comment by NMinhNguyen** [comment]
I'm curious - how can ESM with a default export be added in a backwards compatible way? This has come up before (e.g. https://github.com/facebook/react/pull/18187 which also links to related issues). The problem is with webpack CJS <-> ESM interop where if you have CJS code doing `require('react')` what webpack will return in the presence of an ESM `react` with a default export, is an object with `default` property (meaning it now requires `require('react').default`). But maybe if you also export named then it won't be a problem? I think @TrySound has run into such issues in other packages before.

**Comment by joeldenning** [comment]
> But maybe if you also export named then it won't be a problem?

Yes, this is the approach I am thinking of. See the last 40 lines of https://unpkg.com/browse/@esm-bundle/react@16.13.1/esm/react.development.js - it's an unofficial version of React that does exactly this and is used by several organizations as a drop-in replacement for the official React. No breaking changes.

**Comment by NMinhNguyen** [comment]
> > But maybe if you also export named then it won't be a problem?
> 
> 
> 
> Yes, this is the approach I am thinking of. See the last 40 lines of https://unpkg.com/browse/@esm-bundle/react@16.13.1/esm/react.development.js - it's an unofficial version of React that does exactly this and is used by several organizations as a drop-in replacement for the official React. No breaking changes.

But are you consuming it from CJS code or ESM? Because it's the CJS <-> ESM interop issues that can be very surprising.

**Comment by ljharb** [comment]
@gaearon to be clear; it makes sense to have no default export in the ESM wrapper i'm proposing; anyone doing native ESM would do `import * as React from 'react'` to work around it. However, it's fair that anyone doing that *now* would see it as a breaking change to suddenly not have the default export, so it'd have to wait til v17 if you didn't want to add the default now.

**Comment by joeldenning** [comment]
> But are you consuming it from CJS code or ESM? Because it's the CJS <-> ESM interop issues that can be very surprising.

I've successfully imported it in webpack from both CJS and ESM files.

> anyone doing native ESM would do import * as React from 'react' to work around it. However, it's fair that anyone doing that now would see it as a breaking change to suddenly not have the default export, so it'd have to wait til v17 if you didn't want to add the default now.

Agreed. If starting from scratch, there'd be no need to have the default export. However, without adding a default export, it is not possible to implement [Phase 2](https://github.com/facebook/react/issues/11503#issuecomment-663748887) without it being a breaking change. I personally would be fine with doing Phase 1 in react 16 and Phases 2-4 in React 17+, although my preference is to do Phase 1, 2, and maybe even 3 (with the help of @ljharb's exports checker tool) in React 16 without breaking changes. The reason is that phase 2 is the big one where the majority of users start using the ESM bundle, whereas phase 1 is mostly for early adopters / enthusiasts.

**Comment by dantman** [comment]
Continuing from [this proposal](https://github.com/facebook/react/issues/11503#issuecomment-663748887) on. This seems to be a proposal for a switch to a dual-package with full CJS and ESM source dealing with the dual-package hazard by isolating state somewhere else ([approach 2](https://nodejs.org/api/esm.html#esm_approach_2_isolate_state)). As opposed to using an ESM wrapper like my previous RFC ([approach 1](https://nodejs.org/api/esm.html#esm_approach_1_use_an_es_module_wrapper)).

Assuming that, I do have a few notes of things that have not been mentioned yet.

- You cannot have a Node.js package where both ESM and CommonJS files use `.js`. If you set `"type": "module"` then all of the CommonJS files need to use the `.cjs` file extension instead.
- State and equality dual-package hazard is not the only issue with having both CJS and ESM. Having potentially 2 versions of the same package loaded also increases the memory footprint of React in those cases, and React is not a small library. This is not a deal breaker, but it is worth keeping in mind.
- I do see potential for a dual-package hazard besides React's internal state. For instance if the implementation of the `Component` class is not part of the CJS code where we share state and is instead part of the CJS/ESM bundles then there is risk of `instanceof Component` checks in various libraries breaking.

**Comment by joeldenning** [comment]
> You cannot have a Node.js package where both ESM and CommonJS files use .js. If you set "type": "module" then all of the CommonJS files need to use the .cjs file extension instead.

This is true for NodeJS (but not for bundlers), so the files in the `cjs` directory would have to end with `.cjs`.

> Having potentially 2 versions of the same package loaded also increases the memory footprint of React in those cases

I see how this increases the tarball size published to npm, and thus the overall size on disk. But I don't see how this impacts memory. As far as I know, bundlers and NodeJS do not bring code into memory that has not been loaded via `import` / `require()`. Could you clarify how the memory footprint would change?

> For instance if the implementation of the Component class is not part of the CJS code where we share state and is instead part of the CJS/ESM bundles then there is risk of instanceof Component checks in various libraries breaking.

One proposed solution ([1](https://github.com/facebook/react/issues/11503#issuecomment-663754011), [2](https://github.com/facebook/react/issues/11503#issuecomment-663756703)) is to have the NodeJS esm implementation be simply an ESM interface that calls into the CJS code. That way there is only one definition of `Component` that is ever used in Node. Phase 1 and Phase 2 would not change what runs in NodeJS, though, so this would only apply to Phase 3.

**Comment by sokra** [comment]
As we (webpack) recently also added `exports` field support to webpack 5 I want to give my 2 cents to this topic:

* This work-in-process document has a lot information about the `exports` field regarding webpack but also Node.js and in general: https://gist.github.com/sokra/e032a0f17c1721c71cfced6f14516c62
* These are the keypoints compared to Node.js:
  * webpack 5 also adds `development` and `production` conditions, which are very useful for react. (`process.env.NODE_ENV`, while still supported, should be avoided for frontend code in general, it's Node.js specific)
  * webpack (and other bundlers) supports `require("esm")`, which allows to avoid the dual-state problem by always using ESM (even for `require()`). webpack has introduced a special condition for that: `module`. For this CommonJs and ESM version must export the same interface. Currently there is no other thing that has the dual-state-problem than Node.js. I don't expect that we will see something in future, as this is mostly a backward-compat problem. 
For maximum compatibility I would recommend the following:

#### package.json

``` json
{
    "type": "commonjs",
    "main": "index.js",
    "module": "esm/wrapper.js",
    "exports": {
        ".": {
            "node": {
                "development": {
                    "module": "./esm/index.development.js",
                    "import": "./esm/wrapper.development.js",
                    "require": "./cjs/index.development.js"
                },
                "production": {
                    "module": "./esm/index.production.min.js",
                    "import": "./esm/wrapper.production.min.js",
                    "require": "./cjs/index.production.min.js"
                },
                "import": "./esm/wrapper.js",
                "default": "./cjs/index.js"
            },
            "development": "./esm/index.development.js",
            "production": "./esm/index.production.min.js",
            "default": "./esm/index.production.min.js"
        },
        "./index": "./index.js",
        "./index.js": "./index.js",
        "./umd/react.development": "./umd/react.development.js",
        "./umd/react.development.js": "./umd/react.development.js",
        "./umd/react.production.min": "./umd/react.production.min.js",
        "./umd/react.production.min.js": "./umd/react.production.min.js",
        "./umd/react.profiling.min": "./umd/react.profiling.min.js",
        "./umd/react.profiling.min.js": "./umd/react.profiling.min.js",
        "./package.json": "./package.json"
    }
}
```

#### esm/package.json

Allows to use `.js` as extension in these directory. Alternatively `.mjs` could be used, but this could have potential side-effects when tools check the extension. So `.js` to be safe.

``` js
{
    "type": "module"
}
```

#### esm/wrapper.js

This wrapper is needed for Node.js to avoid the dual-state problem.

``` js
import React from "../cjs/index.js";
export const {
    Children,
    Component,
    ...,
    useState,
    version
} = React;
export { React as default };
```

#### cjs/index.js

This is used by Node.js when `development` and `production` conditions are not supported.

``` js
'use strict';

if (process.env.NODE_ENV === 'production') {
  module.exports = require('./cjs/react.production.min.js');
} else {
  module.exports = require('./cjs/react.development.js');
}
```

#### esm/wrapper.development.js (esm/wrapper.production.min.js similar)

These wrappers are needed for Node.js to avoid the dual-state problem.
They are only used once Node.js adds `development` and `production` conditions.

``` js
import React from "../cjs/index.development.js";
export const {
    Children,
    Component,
    ...,
    useState,
    version
} = React;
export { React as default };
```

#### index.js

For backward-compat.

``` js
module.exports = require('./cjs/index.js');
```

#### esm/index.development.js, esm/index.production.min.js

This is used by tools that does support the `exports` field, `module` condition and `production`/`development` conditions.

``` js
/* React in ESM format */

// compat for the default exports with support for tree-shaking
import * as self from "./esm/index.development.js";
export { self as default }
```

### Results

* webpack 5: `./esm/index.development.js` or `./esm/index.production.min.js`
* browserify: `./cjs/index.js`
* webpack 4 from .mjs: `./cjs/index.js`
* other bundlers: `./esm/wrapper.js`
* Node.js (ESM): `./cjs/index.js` (require) or `./esm/wrapper.js` (import)
* Node.js (old): `./cjs/index.js`
* Node.js (ESM + dev/prod): `./esm/wrapper.development.js` or `./esm/wrapper.production.min.js` for import, `./cjs/index.development.js` or `./cjs/index.production.min.js` for require

### Notes

There is no `esm/index.js` as conditionally choosing a version is not possible in ESM without major trade-offs.
Tools can only benefit fully from react ESM when they support the `exports` field, the `module` condition (because of the dual-state problem) and `production`/`development` conditions (because of the conditional import problem).

Tools can partially benefit from react ESM when they support the `module` field or the `exports` field.

`import { useState } from "react"` or `import * as React from "react"` are technically illegal as long react is an CommonJs module.
Most tooling still support it for backward-compat, but some don't, e. g. Node.js
So currently the only way to use react which is valid everywhere is: `import React from "react"`.
This way should stay supported, otherwise there would be cases (e. g. Node.js 14) where there is no syntax which is valid in react now and react after ESM addition.

Node.js rejected the addition of a `development`/`production` condition for `exports` for now.
That's sad, and I still hope the best that they will eventually add that.
That's why support for that is prepared in the `exports` field above.

**Comment by frehner** [comment]
@sokra great breakdown, very helpful, thank you!

One small question: 

>Node.js rejected the addition of a development/production condition for exports for now.

my understanding is that it's still being worked on? https://github.com/nodejs/node/pull/33171 but maybe I'm misunderstanding that PR

**Comment by joeldenning** [comment]
Thanks @sokra, those are very helpful suggestions.

Here are the options I see. It seems that all of them are technically possible, and that the decision is one of strategy more than of technical implementation:

## Option 1

Add `export default React` to a React 16 ESM build, and remove it once CJS support for `import React from 'react'` is dropped (perhaps in React 17, 18?).

## Option 2

Do not add `export default React`, but create a React 16 ESM build.

## Option 3

Do not publish a React 16 ESM build. (😢) Wait until `import React from 'react';` support is dropped before creating an ESM build.

## Comparison


|  | Option 1  | Option 2  | Option 3  |
| - | -------- | -------- | -------- |
| Unreferenced ESM build | v16 | v16 | v17+ |
| package.json "module" | v16 | v17+ | v17+ |
| package.json "type" / "exports" | v17+ | v17+ | v17+ |

My preference is towards Option 1, as I've explained above. However, Option 2 also is pretty exciting to me. Option 3 is of course less exciting. From what I have gathered in this github issue, we have the technical expertise to make any of these happen (and probably even the labor!).

**Comment by SagnikPradhan** [comment]
Initial reaction to this issue was, why is this issue open even after **3 years**? After reading through part of it, makes sense why it is taking so long. Maintaining a library like React is a huge task. So 🙇🏻

**Comment by joeldenning** [comment]
Given the recent news with React 17, I have updated my previous comment to reference React 17 instead of 16 for any future plans.

I'd appreciate feedback from people on which of the three above options is preferred.

**Comment by ljharb** [comment]
The `"./": "./"` makes it safe, yes, but also prevents any encapsulation, so you'd want to remove that as soon as you had a semver-major.

**Comment by nstepien** [comment]
FWIW babel outputs the new jsx runtime import as
```js
import { jsxs, jsx, Fragment } from 'react/jsx-runtime';
```
but if you load a module like that in Node, it'll complain about the lack of file extension:
```
> node .\node.mjs
node:internal/process/esm_loader:74
    internalBinding('errors').triggerUncaughtException(
                              ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find module 'test\node_modules\react\jsx-runtime' imported from test\node_modules\react-data-grid\lib\bundle.js
Did you mean to import react/jsx-runtime.js?
    at new NodeError (node:internal/errors:259:15)
    at finalizeResolution (node:internal/modules/esm/resolve:307:11)
    at moduleResolve (node:internal/modules/esm/resolve:742:10)
    at Loader.defaultResolve [as _resolve] (node:internal/modules/esm/resolve:853:11)
    at Loader.resolve (node:internal/modules/esm/loader:85:40)
    at Loader.getModuleJob (node:internal/modules/esm/loader:229:28)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:51:40)
    at link (node:internal/modules/esm/module_job:50:36) {
  code: 'ERR_MODULE_NOT_FOUND'
}
```

Adding `exports` should fix it 🤔

**Comment by Andarist** [comment]
@nstepien providing a full `exports` map as you have shown in your previous post is not an option from what I believe. What node implements regarding cjs interop and stuff doesn't really play well with the existing ecosystem. Dual package hazard is real - especially for packages like Reac that requires a single copy of them.

An `exports` map with commonjs only files could potentially be added without breaking anything but would also have to be done extra carefully and with appropriate e2e tests for this (given how complex things are to get right)

**Comment by ljharb** [comment]
@Andarist it works fine, and is no different in react’s case, which has always had that hazard and the ecosystem solves it by making react a peer dep everywhere. An “exports” map can work just fine here, as long as the ESM files and CJS files share the same state - which can be achieved by writing simple ESM wrappers.

**Comment by Andarist** [comment]
If all dependencies, transitive or not, of React are under the control of React (which in this case they are) and ESM entry points of all of them are just reexporting CJS content then ye - maybe that is achievable in this particular case.

There is still the whole drama of what the actual shape of ESM entry should be though (named, default, both):
- named-only: not really backward-compatible because a lot of the code out there is using `import React from 'react'`, which is also the only way to actually import React in node right now when using ESM
- default-only: not really backward-compatible because a lot of the code our there is using `import * as React from 'react'`, this has often been promoted by type-checkers and other tools
- both: the only way to make it fully backward-compatible, so it can work with all current loading styles and when mixing ESM & CJS modules across the dependency tree

I constantly forget about the possibility of ESM wrappers as they feel like a cheat but also because this technique **only** works if you control all of your dependencies and can't really be used as a universal strategy that would "just work" 😢

**Comment by ljharb** [comment]
I'm afraid the only universal strategy to provide both is, in fact, having all your actual code in CJS, and writing ESM wrappers around it to provide named exports. Code that is not stateful nor relies on identity can be natively *written* in both, but that's a subset, a caveat.

**Comment by nstepien** [comment]
`jsx-runtime` isn't stateful right? Should be safe to ship both esm/cjs with no wrapper for it.

Should I log a separate issue for importing `react/jsx-runtime` in esm node?

**Comment by sokra** [comment]
For frontend package an extra challenge joined the dual state hazard: let call it the double bundle hazard.

When exporting ESM and CJS versions and the package is used via `require` and `import`, both versions would be bundled and double the effective bundle size for the package.

**Comment by nstepien** [comment]
@sokra is that possible in practice? For example with rollup I would assume that since cjs modules are transformed into esm modules, that it would then prefer importing the esm modules whenever available.

**Comment by Andarist** [comment]
Its possible in practice for bundlers following node semantics - such as webpack 5. Its important to match semantics because otherwise u would potentially experience different results when running in node and when running a bundles code.

However, webpack 5 handles a special `"module"` condition that is used to deduplicate esm/cjs (the ones comming from exports map).

If we consider @ljharb’s proposal this isnt really relevant though because he proposes that esm file could be just a few lines of wrapper code that would just reexport the cjs file.

**Comment by ljharb** [comment]
With a thin ESM wrapper, there is zero additional hazard, with or without bundlers - only the normal one that peer deps avoid about dupes in the graph.

**Comment by juliuskovac** [comment]
With all respect, wrapper doesn't sound like cheating, it is cheating. Tree shaking and other analysis tools which know only about ES6 modules semantics won't fully work with it ever. If there will be major release of react, I think it will be more beneficial for community to go with correct ES6 modules and named-only exports, that's what versioning is about to do breaking changes if they are really necessary. If there is problem with tooling (to bundle two versions or other issue) it should be probably fixed on tooling side.

**Comment by ljharb** [comment]
@juliuskovac tree shaking works equally well with CJS, it's just not enabled by default in the popular bundlers, so they could *already* handle things without any ESM whatsoever if it was important to do so. The problem that needs fixing on the tooling side is "treeshaking often only works with ESM by default".

**Comment by lukastaegert** [comment]
I assure you, it is not a question of "defaults". It is not impossible, but at least for Rollup, there is still some engineering effort necessary to safely track usage of object properties as would be required for CommonJS tree-shaking. And beyond tree-shaking, scope-hoisting is another optimization you forego without ES modules. In short it means you have a lot more property accesses with the incurred runtime penalty in a CommonJS setting. Rollup is actively trying to reduce these, but it requires libraries to be sufficiently "well-behaved". In short, it is quite hard to produce similar quality output from CommonJS modules as opposed to ES modules.

**Comment by ljharb** [comment]
Since dynamic imports exist, well-behaved libraries are required anyways.

**Comment by hronro** [comment]
The [plan for React 18](https://reactjs.org/blog/2021/06/08/the-plan-for-react-18.html) is finally announced, but I'm a little disappointed that there is no mention of ES module support at all.
ES module support is a feature request for a long time, and even this issue has been opened for about 4 years, I really hope the new major release (React 18) could have the feature implemented.

**Comment by bvaughn** [comment]
One of our main goals for the 18.0 release is to remove as many upgrade obstacles as possible so that people can migrate and began taking advantage of the new features and fixes. ES modules would be a breaking change (so they would compete with this goal). They are still _important_ to us though, and we hope to make the change sometime in the near future, just not as part of 18. :smile:

**Comment by wereHamster** [comment]
There is currently much confusion about whether to `import React from` or `import * as React from`. Both work, currently, but that's only accidental due to how ESM imports of CommonJS modules are treated by bundlers.

I'd really appreciate an official stance by the React developers as to which form will be supported once React starts providing native ESM, so that we can start educating people now, update examples, blog posts, provide linter rules, emit bundler warnings etc now, rather than wait with it until the ESM support finally ships…

**Comment by milesj** [comment]
Pretty sure the path forward is named imports only, so `import { useState } from 'react';`. The default import is _no longer required_ if you are using the new automatic JSX transform.

**Comment by bvaughn** [comment]
@milesj is correct. The "React" import is no longer required for JSX usage, thanks to the efforts of @lunaruan last year (via #18484, #18299, #18023, etc)

**Comment by hronro** [comment]
I'm a little bit confused about the roadmap now. If ES module will not come with v18.0, will it come with v18.x or v19.0? (or ... even later?)

I'm not sure if React.js follow semantic versioning. If it is, it's actually OK to introduce a break change in a major release.

If React.js doesn't follow semantic versioning and finally support ES module in v18.x, it's still weird that React.js does not introduce break changes in major version but introduce them in a minor version.

If React.js finally support ES module in v19.0, OK, maybe we have to wait another 4 years for the ES module support.

**Comment by thientran1707** [comment]
may I know if we have any status for the ESM module support from React?

**Comment by GabrielDelepine** [comment]
> When would an official ESM version of React be finally published?

If the transition to ESM is not made properly, it will be a terrible mess. I am like you, I can't wait but I definitely prefer the team to take enough time to ship something great. Don't underestimate the amount of work, it's huge because they are so many implications

Until then, I am loading `react` and `ReactDOM` as a UMD global in my ESM repos and it's doing the job. Everything stays typed out of the box

package.json
```
{
  "type": "module"
}
```

tsconfig.json (Typescript 5.0)
```
{
  "compilerOptions": {
    "allowUmdGlobalAccess": true,
    "jsx": "react",
    "module": "nodenext",
    "moduleResolution": "nodenext",
    "resolveJsonModule": true,
    "strict": true,
    "target": "ES2022",
    "verbatimModuleSyntax": true
  }
}
```

index.html
```
    <script type="importmap">
      {
        "imports": {
               <====== list your ESM dependencies here
        }
      }
    </script>
    <!-- of course, only expose the desired files on production, not your complete node_modules/ folder -->
    <script src="../node_modules/react/umd/react.development.js"></script>
    <script src="../node_modules/react-dom/umd/react-dom.development.js"></script>
    <script type="module" src="./index.js"></script>
```

<img width="1047" alt="Screenshot 2023-04-28 at 19 08 45" src="https://user-images.githubusercontent.com/1979087/235210710-1e8790ae-0d60-41af-8484-be275cfb070f.png">

**Comment by xiaoqiang1999** [comment]
Please support ESM as soon as possible, it is useful, thanks.
For example, Vue.js supported ESM long ago.

It is currently available for temporary use [https://esm.sh/](https://esm.sh/) To get the ESM package for React.

**Comment by Starcounter-Jack** [comment]
I second the suggestion to use `https://esm.sh/react@18.2.0`  as they have provided a proper modern ESM distribution of React for a very long time.

**Comment by jonkoops** [comment]
> So our current thinking is that when we go ESM, we might want to try go ESM all the way. No CJS at all — or separated in a compat legacy package. This won't happen in React 17 and is unlikely in 18, but is plausible to try in React 19.

@gaearon Now that the new JSX transform has been out for a while, there should not longer be a blocker right? Perhaps the time is right to do this with React 19. How would you feel if I drafted up a renewed proposal along the lines of https://github.com/reactjs/rfcs/pull/38?

**Comment by luzzif** [comment]
Guys please, is there any update on this?

**Comment by alshdavid** [comment]
> Bundling is for optimisation [..] IMHO, a bundler should no longer be a development time dependency in 2023.

There will always be need for a build step for sources as shipping sources directly cannot be as efficient as a tool designed to organize sources into an optimized output. Bundles are reasonable as they can take advantage of compression, dead code elimination, and bundlers/build systems can help handle consuming invalid sources (like consuming TypeScript, importing css, importing images, markdown, etc).

>  should provide compatibility with ESM and not fight against it.

React using CJS causes. bundlers to de-opt its output. An example of such a de-opt is that - in the case of an ESM export like this:
```javascript
export const createElement = () => {}
```
A bundler knows this value cannot be reassigned and therefore generates a direct property assignment within the isolated module scope
```javascript
modules['react'] = (bundler_require, module) => {
  const createElement = () => {}
  module.createElement = createElement
}
```
However with CJS, a property assignments can be done internally within the module, externally by other modules, synchronously within a nested scope, etc - this means we can never guarantee that the value remains unchanged.
```javascript
module.exports.foo = 'something'

module.exports.bar = function() {
  module.exports.foo = 'something else' 
}
```
So bundlers have to use getters/setters to "proxy" onto cjs properties
```javascript
modules['react'] = (bundler_require, module) => {
  let foo_unique_id_1 = 'something'
  let bar_unique_id_2 = function() {
    foo_unique_id_1 = 'something else' 
  }
  Object.defineProperty(module, 'foo', { get: () => foo_unique_id_1, set: (v) => foo_unique_id_1 = v })
  Object.defineProperty(module, 'bar', { get: () => bar_unique_id_2, set: (v) => bar_unique_id_2 = v })

  // Synthetic default exports
  Object.defineProperty(module.default, 'foo', { get: () => foo_unique_id_1, set: (v) => foo_unique_id_1 = v })
  Object.defineProperty(module.default, 'bar', { get: () => bar_unique_id_2, set: (v) => bar_unique_id_2 = v })
}
```
Getters are 10x slower than direct property accesses and that adds up in a big project with lots of React in it.

Please distribute a copy of React as unminified ESM and let the bundlers do the heavy lifting of optimizing.

**Comment by hronro** [comment]
What a shame that there is still no ESM support in the final React.js v19.0 release, or not even mention it in the [release blog post](https://react.dev/blog/2024/12/05/react-19).

**Comment by jonkoops** [comment]
My guess would be that making the new JSX transform mandatory was a first step towards this, so perhaps it will be React v20 that will enforce ESM? Still, kind of sad to not see at least a formal plan or acknowledgement since this is in the top 10 [most upvoted issues](https://github.com/facebook/react/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by spamshaker** [comment]
this is really one liner to build it for production - maybe this way it could be exposed temporarly

 `bun build --minify --outfile=./dist/react.mjs ./node_modules/react/cjs/react.production.js`
and adding "module"="./dist/react.mjs" to a package.json

same goes for other packages

**Comment by spamshaker** [comment]
I'm sorry I can't wait longer. I will switch to the target solution when this issue will be resolved, but now I dealt with it on my own => https://github.com/ria-develop/react-esm
```html
<!doctype html>
<html lang="en-US">
<head>
  <meta charset="UTF-8">
  <title>Example of React ESM use case</title>
  <script type="importmap">
    {
      "imports": {
        "react/jsx-runtime": "https://cdn.jsdelivr.net/npm/@ria-develop/react-esm",
        "react": "https://cdn.jsdelivr.net/npm/@ria-develop/react-esm",
        "react-dom": "https://cdn.jsdelivr.net/npm/@ria-develop/react-esm",
        "react-dom/client": "https://cdn.jsdelivr.net/npm/@ria-develop/react-esm"
      }
    }
  </script>
  <script type="module">
    import { createRoot } from "react-dom/client";
    import { StrictMode, Suspense } from "react";
    import { jsx, jsxs } from "react/jsx-runtime";

    function App() {
      return jsxs(Suspense, { fallback: "Loading...", children: ['hello'] });
    }

    const root = createRoot(document.getElementById("root"));
    root.render(jsx(StrictMode, { children: jsx(App, {}) }));
  </script>
</head>
<body>
<div id="root"></div>
</body>
</html>
```
till then, I am going to prepare a gitlab action which will rebuild it and sync versions with react releases

**Comment by aladdin-add** [comment]
@spamshaker you don't have to do this - there are several cnds can serve for this, e.g. https://esm.sh/:

```js
import React from "https://esm.sh/react"
```

**Comment by alshdavid** [comment]
Or just use Preact

**Comment by Hexagon** [comment]
Are we there yet? 👀

**Comment by codyfrisch** [comment]
Aaand this is why we bit the bullet and left the Recrapct ecosystem. Couldn't be happier.

**Comment by armano2** [comment]
> My impression is that people who would import `*` (and not use the default) wouldn't have problems tree shaking since default would stay unused. But maybe I overestimate Rollup etc.

there is actually preety good reason to allow users to import esm directly, mixing esm and cjs is costly, and most bundlers do convert this at runtime, eg. 

```js
// ------- Convert cjs to esm-------
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __commonJSMin = (cb, mod) => () => (mod || cb((mod = { exports: {} }).exports, mod), mod.exports);
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") {
		for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
			key = keys[i];
			if (!__hasOwnProp.call(to, key) && key !== except) {
				__defProp(to, key, {
					get: ((k) => from[k]).bind(null, key),
					enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
				});
			}
		}
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));

// ------- import -------

var require_react_production = /* @__PURE__ */ __commonJSMin(((exports) => {
 /// ..... react code
}));
var require_react = /* @__PURE__ */ __commonJSMin(((exports, module) => {
	module.exports = require_react_production();
}));
var import_react = /* @__PURE__ */ __toESM(require_react(), 1);

/// your component

import_react.createElement("span", ...);
```

this is really hard to optimize, and basically not treeshaken at all if you mix cjs + esm

with esm exports, functions eg `createContext` could be optimized away if they are not used


-----------------------------

bdw, you can see this clearly in react.dev 
```
      t.unstable_useCacheRefresh = function () {
        return E.H.useCacheRefresh()
      },
```
its never referenced, and can't be removed https://react.dev/_next/static/chunks/framework-76ad44d61c11f057.js

-----------------------------

`import *` .. is preety trivial to tree-shake compared to cjs

**Comment by jonkoops** [comment]
It seems like the question in this issue's title (what the top-level ES exports should look like) was settled internally years ago:

- @gaearon's [update on this issue](https://github.com/facebook/react/issues/11503#issuecomment-662647468) (Jul 2020): *"Our eventual plan is to move away from default exports entirely. [...] `import React from 'react'; // no`"*
- @gaearon on [RFC #38](https://github.com/reactjs/rfcs/pull/38#issuecomment-901360125) (Aug 2021): *"Our latest thinking was that we'd drop default exports altogether. Named only."*
- @sebmarkbage in the description of [#18102](https://github.com/facebook/react/pull/18102) (Feb 2020): *"the ES module will not have a default export. Only named exports."*
- @bvaughn [confirmed](https://github.com/facebook/react/issues/11503#issuecomment-859741899) the path forward is named imports (Jun 2021)
- The automatic JSX transform (now mandatory in React 19) removed the main dependency on `import React from 'react'`

But this was never formally documented as a decision on this issue, and it's blocking downstream tooling from acting. For example, eslint-plugin-react has had a [fully reviewed, maintainer-approved PR](https://github.com/jsx-eslint/eslint-plugin-react/pull/2585) for a `named-import` rule sitting in limbo since 2021, partly because there's no official public statement to point to.

Would it be possible to get a definitive comment here confirming that React's future ESM entry points will use named exports only, with no default export? That alone, even without a timeline for actually shipping ESM, would let the ecosystem start preparing: lint rules, codemods, documentation, and gradual migration.

If it would help move things forward, I'd be happy to contribute a formal RFC building on [RFC #38](https://github.com/reactjs/rfcs/pull/38) that reflects the current state of the ecosystem (React 19's mandatory JSX transform, Node's mature ESM support, the `package.json` `exports` field) and lays out a concrete migration path.
